### ISSUE TITLE
Manager looks for the actual bro scripts only from the top level dir in the repo

### ISSUE THREAD

**Issue by mkayoh** [bro/package-manager#2]
I was trying to set up a repo of my own to work with the package manager. I had two directories, one for index files and the other for the actual bro scripts. Both directories had as well sub directories for categories and inside them the actual index files/bro scripts. The search function worked just fine but the process failed when trying to download the scripts with the following error message.

`2016-10-13 07:09:14 INFO     getting info on "angler_goon": invalid git repo path: 'git clone --depth=1 https://github.com/mkayoh/bro-scripts/scripts/intel/angler_goon /home/ubuntu/.bro-pkg/scratch/angler_goon' returned with exit code 128
stderr: 'Cloning into '/home/ubuntu/.bro-pkg/scratch/angler_goon'...`
`fatal: repository 'https://github.com/mkayoh/bro-scripts/scripts/intel/angler_goon/' not found'
"bro/packages/intel/angler_goon/angler_goon" info:
        invalid package: package is no longer a valid or reachable git repo`

Meaning that right now I'd need one repo per bro script I want to download with the package manager.

**Comment by jsiwek** [comment]
Right, so I think there's two things to clarify about how the package manager currently works:

1. Each "package" has its own git repository.
2. A "package source" has its own dedicated git repository and contains the bro-pkg.index files that contain metadata which point out the git URL for packages.

The intention of each package having its own git repo is so that unrelated Bro scripts/plugins don't have to share versioning and dependency information.  i.e. git is providing the versioning information, so package "foo" can be at 1.0 and package "bar can be at 2.0 and users can download/install/upgrade them separately and any change in "bar" does not affect "foo".

So ideally, I'd suggest trying to identify the smallest chunks of functionality that a user would want to be able to install on its own and make each of those a package.  However, you could also start with a single monolithic package full of all your scripts and decide to make it more modular over time or whenever its called for.

Does that help?

**Comment by mkayoh** [comment]
First of all I love the idea of the package manager but as it stands the design is terribly wasteful. I have currently dozens of different Bro rules and I'd like to be able to install/uninstall them individually easily. Some of them interlace so I wouldn't even want them installed at the same time. Think of it this way, if I install a package on Debian with apt-get I'm usually faced with a question if I want to install some additional packages that are **needed** by the package I actually want to install. They're necessary for the functionality of the main package, they're not optional. The design choice you've here would force me to cut corners and come up with hacky solutions trying to group up somewhat similar packages which is not in any way [ideal for the user](https://en.wikipedia.org/wiki/KISS_principle) installing the packages.

If the intention of the package manager is actually to install Bro scripts in bundles I can't quite frankly see the point of it. I could just as easily write a few lines bash script that clones the git repo to the Bro script/plugin directory. Besides no matter how well the Bro scripts are grouped they're going to need loads of repos and as it stands I think that goes against how Github was designed. Granted it does give the option to keep track of the versioning but what then is the point of the package version in the bro-pkg.meta files?

What I would suggest is to change the functionality to clone the entire repo and afterwards move the packages from there to the Bro scripts/plugins dirs. The bro-pkg.meta files can still be used to keep track of versions as well as dependencies. The latter is needed in any case since there's no function in Github to keep track of the various depencies for bro scripts/plugins and it's just easier for the devs themselves to make sure they're providing that information when sharing their own scripts/plugins with others.

**Comment by jsiwek** [comment]
> First of all I love the idea of the package manager but as it stands the design is terribly wasteful. I have currently dozens of different Bro rules and I'd like to be able to install/uninstall them individually easily. Some of them interlace so I wouldn't even want them installed at the same time. Think of it this way, if I install a package on Debian with apt-get I'm usually faced with a question if I want to install some additional packages that are needed by the package I actually want to install. They're necessary for the functionality of the main package, they're not optional. The design choice you've here would force me to cut corners and come up with hacky solutions trying to group up somewhat similar packages which is not in any way ideal for the user installing the packages.
> 
> If the intention of the package manager is actually to install Bro scripts in bundles I can't quite frankly see the point of it. I could just as easily write a few lines bash script that clones the git repo to the Bro script/plugin directory.

No, it's not the intention for unrelated scripts to be bundled together in monolithic packages.  But because I didn't fully understand the specifics of what you're trying to do, I was just suggesting that the package manager could technically be used that way on the chance that it would be a suitable option for you or just a temporary workaround.

The intention is for bro-pkg to be more similar to package managers like apt-get, like you mentioned.  e.g. it would auto-install required dependencies (but automatic dependency analysis/resolution isn't implemented in bro-pkg yet)

> Besides no matter how well the Bro scripts are grouped they're going to need loads of repos and as it stands I think that goes against how Github was designed.

I'm not sure I understand the concern, can you elaborate?

> Granted it does give the option to keep track of the versioning but what then is the point of the package version in the bro-pkg.meta files?

That's a good point.  The 'version' key in bro-pkg.meta I think is just a relict from previous design and no longer serves a purpose, so I actually plan to remove it now.  Git will be the authority on package version information.

> What I would suggest is to change the functionality to clone the entire repo and afterwards move the packages from there to the Bro scripts/plugins dirs.

I don't understand the suggestion because at the moment there's no mechanism to group multiple packages within a single repo (sorry if my earlier suggestion confused you).  Each package is its own git repository.

Then there is another thing called a "package source", which is a separate type of git repo and has a collection of index files which point at the git URLs of packages.   Since you seem familiar with Debian package management, this is analogous to a "repository" or "software channel" in the APT package manager.

Maybe looking at the organization of https://github.com/bro/packages as an example would help?

> The bro-pkg.meta files can still be used to keep track of versions as well as dependencies.

I don't think keeping track of versions in bro-pkg.meta is adequate. e.g. lets say package "foo" has version 1.5 available, but I want to install version 1.3.  How does bro-pkg actually get version 1.3 if that info is only available in the bro-pkg.meta file?  Does it search through all historical git commits until it finds a version of bro-pkg.meta that says 1.3 ?  What if the developer forgot to change that information in bro-pkg.meta and so multiple commits say 1.3 ?  This is why git tags are used as the authority on package versioning and it makes sense to use git for versioning -- it is a version control system after all.

But yes, dependency information will go in bro-pkg.meta once it's implemented.

Hope that helps explain the design decisions.   I think it is designed in a way that will work similar to how you expect (e.g. like apt-get) but either my previous explanation/suggestion was confusing, or I don't understand what you're trying to do.  You are trying to set up your own "package source" repository, right?

**Comment by mkayoh** [comment]
> I'm not sure I understand the concern, can you elaborate?

So here I have my own [repo](https://github.com/mkayoh/bro-scripts) as an example. I currently have there 44 different rules that I want to install individually. Some of them might have some similar functionalities, for example I could group all the packages that are looking for XSS in one place or the other and make the number smaller. However, this would create a problem with users who might not need for example typosquatting SMTP since there is no email traffic in their network so this script would be redundant for them. This is why I want all my scripts served individually. This on the other hand would mean that currently I'd have to create 44 new repos on my account. I could create a dump account just for these repos except that it goes against Github's [terms of service](https://help.github.com/articles/github-terms-of-service/). So my only option would be to create 44 repos on my own account which would clutter my own account with repos that contain barely three or four different files.

> I don't understand the suggestion because at the moment there's no mechanism to group multiple packages within a single repo (sorry if my earlier suggestion confused you). Each package is its own git repository.

Yes, I know this and this is in my mind how you should restructure the package manager.

> I don't think keeping track of versions in bro-pkg.meta is adequate. e.g. lets say package "foo" has version 1.5 available, but I want to install version 1.3. How does bro-pkg actually get version 1.3 if that info is only available in the bro-pkg.meta file? Does it search through all historical git commits until it finds a version of bro-pkg.meta that says 1.3 ? What if the developer forgot to change that information in bro-pkg.meta and so multiple commits say 1.3 ? This is why git tags are used as the authority on package versioning and it makes sense to use git for versioning -- it is a version control system after all.

I understand your argument with the versioning via Git tags but it still doesn't justify in my mind the requirement for each package to have its own repo. Granted it does have its merits over the developers themselves keeping track of the of versioning in the bro-meta.pkg files but at the same time it isn't the best solution either. Yes Git tags were designed exactly the version control in my mind but Github wasn't designed to be package repository. I would suggest writing an additional functionality to bind the versioning to commits in the directories containing the packages. This just my first thought and is far from optimal but since Github wasn't designed for this kind of use-case scenario you will ultimately have to choose whether to have the versioning via Git tags which are a sure way or whether to have a better experience for the user. Optionally it might be a better idea to use some other service for hosting the packages.

Let's have a little thought experiment. I have no statistics on how many public repos exactly on Github host bro scripts but with a simple search for the word bro there are 40, 307 different repositories. Some of them are false hits and are linked to something else like browser so let say 20k of those are Bro related. Now most of those repos are forks from different repos so let's assume 1/10 of those are unique repos. That would make 2k repos which is still a big number and I'd wager a guess after filtering out the repos containing something other than Bro scripts the real figure is under 1k. So if those 1,000 repos each contain let's say 4 packages which is a generous estimate as most contain way more, this would mean that there would have to be four times this amount of new repositories to contain all the packages and those 4,000 would in all likelyhood contain only a couple of files. This is not only a wasteful practice but also goes against the idea of Github which is to host code for projects, not as a personal data storage. The numbers are just guesstimates but the point stands that hosting the packages can be done better in Github by either allowing multiple packages in a single repo or hosting them some other place than Github.

**Comment by jsiwek** [comment]
> So my only option would be to create 44 repos on my own account which would clutter my own account with repos that contain barely three or four different files.

Ok, I'm seeing now what you're concern is about, but to help me understand better: is it just with respect to your perception of "clutter" or are there usability issues that creating that many repos would cause for a user?

> I would suggest writing an additional functionality to bind the versioning to commits in the directories containing the packages. This just my first thought and is far from optimal 

Yeah, while I see something like that as technically possible, it seems like it would be such an out-of-the-ordinary way of using git that it would defeat the purpose of trying to make things easy/familiar for developers.
 
> mean that there would have to be four times this amount of new repositories to contain all the packages and those 4,000 would in all likelyhood contain only a couple of files.

If we're just talking about the total number of packages, 4k would still seem in line with the number of packages in similar package management solutions.  E.g. Rust's cargo/crates has ~6.5k packages and Homebrew has ~3.5k formula/packages.

> This is not only a wasteful practice

Can you explain that more, I'm not sure what is being wasted?

With respect to GitHub, I don't think there's a limit on number of repos you can have, is there something else that makes many repos problematic/wasteful?

> but also goes against the idea of Github which is to host code for projects, not as a personal data storage.

So even if you end up creating a git repo for a package containing a single script, does that not still fall under the realm of "hosting code for projects" ?  Just because it's a single file/script doesn't make it any less of a project or less deserving of its own independent version control.

Or if by "personal data storage" you mean things that aren't really meant to be useful to others except yourself, I'd say even that is not an unusual use of GitHub.  E.g. people storing their personal dotfiles on GitHub.

> the point stands that hosting the packages can be done better in Github by either allowing multiple packages in a single repo

From your [example repo] (https://github.com/mkayoh/bro-scripts) you're trying to set up, I can see how multiple packages in a single repo would make your life easier, but I'm also not sure your use case is going to be common or even needed as bro-pkg matures and gets used by more people.  By that I mean it looks like many of the scripts in your repo are collected from various other sources/authors and so in an ideal package ecosystem those authors would be be maintaining the individual packages themselves and you'd just be consuming them.

The approach we took with bro-pkg was to try to keep things as decentralized as possible, which is at odds with your centralized approach.  But maybe that's "ok" if the only real problem you have is that there will be a lag time before the original package authors adopt the packaging model...

So let me ask a hypothetical: if all the original authors of all packages in your example repo already had set up packages and registered with bro-pkg's default package source (i.e. on a fresh bro-pkg install, you could `bro-pkg install X` for any X you can imagine) would you still be motivated to set up and organize a repository similar to [your example] (https://github.com/mkayoh/bro-scripts) ?

> or hosting them some other place than Github.

You should already be able to host your package repos anywhere you want (e.g. local filesytem, private server, Bitbucket, GitHub, etc.) and bro-pkg will still work.

**Comment by mkayoh** [comment]
> Ok, I'm seeing now what you're concern is about, but to help me understand better: is it just with respect to your perception of "clutter" or are there usability issues that creating that many repos would cause for a user?

My main concern is the disorganization it creates. Yes I could do it and it would take time but I could do it. Afterwards I'd have all those forty repos on my account which I do not want. Sure I would have working versioning on each and every one of them but I would also fill my account with repos that will in all likelyhood get attention once a year each.

> By that I mean it looks like many of the scripts in your repo are collected from various other sources/authors and so in an ideal package ecosystem those authors would be be maintaining the individual packages themselves and you'd just be consuming them.

Another concern are organizations. The rules I have in my repo are open for all to use but the company I work for and I'm sure many other companies that use Bro host their own Bro rules as well just for their own use. I can't disclose the number of the rules we don't release publically but it can be compared to my public repo. So herein lies the dilemma, companies usually have way more repos than single users so in my mind it's not reasonable to add to the mix with the additional repos.

As to the maintenance of the packages I frankly don't share your optimism. From the 44 different rules in my example repo only the ones shared by Bro or its core devs were actively maintained. The rest were functional and had really good ideas but I had to go through almost all of them to fix the deprecated functions, dependencies etc. 

> So let me ask a hypothetical: if all the original authors of all packages in your example repo already had set up packages and registered with bro-pkg's default package source (i.e. on a fresh bro-pkg install, you could bro-pkg install X for any X you can imagine) would you still be motivated to set up and organize a repository similar to your example ?

Even as the package manager matures I'm sure the companies that would utilize it would want to host the packages they use themselves for a number of reasons, number one being that in a decentralized solution a company trying to secure a network would have to trust external sources for the Bro rules. A good example on this is the cheap VPS provider DigitalOcean which serves all of the packages for the VMs from its own servers. As you already state in the package manager repo:

> It's left up to users to decide for themselves -- whether any given package is trustworthy as there is no implied guarantees that it's secure just because it's been accepted into the default package source.

So to summarize it up decentralization in security is never a good idea nor is trusting external sources of data. The structure I'm using in my own repo is the one I would propose you to adapt as well. I see the versioning via Git tags is an important argument for you but based on the number of unmaintained Bro script repos I doubt it will have much use if there are no packages in the first place to serve. The thought of filling my account with repos that are really not projects of mine but packages for only one of the projects I participate in isn't an appealing thought and I'm sure I'm not alone in this.

**Comment by jsiwek** [comment]
> As to the maintenance of the packages I frankly don't share your optimism. From the 44 different rules in my example repo only the ones shared by Bro or its core devs were actively maintained. The rest were functional and had really good ideas but I had to go through almost all of them to fix the deprecated functions, dependencies etc.

I think the idea was the strength of the open-source, decentralized model would help prevent some of that bit-rot since it makes it easy to contribute/accept patches.  But yeah, it may be overly optimistic since it's entirely dependent on having an active community.

> So to summarize it up decentralization in security is never a good idea nor is trusting external sources of data. The structure I'm using in my own repo is the one I would propose you to adapt as well.

Ok, thanks for the explanations.  I'll have to think more on how it could support both use-cases and maybe ask others to weigh in with opinions/ideas on what changes to do.

> I see the versioning via Git tags is an important argument for you

Yea,  it is important for the current model of how I thought versioning and inter-package dependencies would work.

But for your model, would it be right to say that per-package versioning is not so important?  i.e. for any given commit/snapshot of your repo, all packages and any inter-package dependencies are in a working state?

If that were true, I think supporting your model would become much simpler rather than if the package manager had to pick and choose packages from various commits in the repo in order to get versioning and dependencies to all work out correctly.

> The thought of filling my account with repos that are really not projects of mine but packages for only one of the projects I participate in isn't an appealing thought and I'm sure I'm not alone in this.

Just checking because I didn't see you respond directly to it even when talking about companies wanting to host their own packages, but you saw that you aren't required to use your GitHub account to host packages?  There's also no need to permanently trust external sources as you could fork/clone into an internal git server after doing a code review/audit and from that point on it's entirely under your control. 

This is kind of the way I expected such cases to be handled since that's a standard way of using git as a distributed VCS.  It has the benefit of making it easy to pull in upstream changes.  But because your model is combining code from multiple upstream sources into a single repo, I think you lose that strength of using git as a distributed VCS: it's now maybe not as simple as it used to be to pull in upstream changes from the original authors?

**Comment by mkayoh** [comment]
> But for your model, would it be right to say that per-package versioning is not so important? i.e. for any given commit/snapshot of your repo, all packages and any inter-package dependencies are in a working state?

That is true.

> Just checking because I didn't see you respond directly to it even when talking about companies wanting to host their own packages, but you saw that you aren't required to use your GitHub account to host packages? There's also no need to permanently trust external sources as you could fork/clone into an internal git server after doing a code review/audit and from that point on it's entirely under your control.

I did see it but the issue I had was mainly with the main design choice of the package manager. I could clone into an internal git server but that would create additional work for updating the packages there.

>  But because your model is combining code from multiple upstream sources into a single repo, I think you lose that strength of using git as a distributed VCS: it's now maybe not as simple as it used to be to pull in upstream changes from the original authors?

That would be true if the packages were themselves actively maintained. Right now even though I'm not the author of the most of the rules I took care of putting them into a working shape. Your model would work if the packages were actually actively maintained but since Bro is currently moving forward at a pretty fast pace and features get deprecated and new ones get added all the time it would be simpler to opt for this kind of model. I have nothing against the package authors themselves taking care of fixing the packages, I just find it very unlikely to happen as is the case with Bro scripts currently.

**Comment by jsiwek** [comment]
There's no plans to alter the core design of bro-pkg.  Ideas for making it easier to accommodate your use-case while still following the distributed dev model it's been designed for are welcome.
